core/src/main/java/org/infinispan/container/entries/CacheEntrySizeCalculator.java

```diff
@@ -27,76 +30,81 @@ public CacheEntrySizeCalculator(EntrySizeCalculator<? super K, ? super V> calcul
-      iceSize += OBJECT_SIZE + POINTER_SIZE;
-      // Each ICE references key and value
-      iceSize += 2 * POINTER_SIZE;
-      boolean mortalEntry;
-      boolean transientEntry;
+      // This will be non zero when use expiration, but don't want to store the metadata
+      long noMetadataSize = 0;
       boolean metadataAware;
       // We want to put immortal entries first as they are very common.  Also MetadataImmortalCacheEntry extends
       // ImmortalCacheEntry so it has to come before
       if (ice instanceof MetadataImmortalCacheEntry) {
-         mortalEntry = false;
-         transientEntry = false;
          metadataAware = true;
       } else if (ice instanceof ImmortalCacheEntry) {
-         mortalEntry = false;
-         transientEntry = false;
          metadataAware = false;
       } else if (ice instanceof MortalCacheEntry) {
-         mortalEntry = true;
-         transientEntry = false;
+         noMetadataSize += 16;
          metadataAware = false;
       } else if (ice instanceof TransientCacheEntry) {
-         mortalEntry = false;
-         transientEntry = true;
+         noMetadataSize += 16;
          metadataAware = false;
       } else if (ice instanceof TransientMortalCacheEntry) {
-         mortalEntry = true;
-         transientEntry = true;
+         noMetadataSize += 32;
          metadataAware = false;
       } else if (ice instanceof MetadataMortalCacheEntry) {
-         mortalEntry = true;
-         transientEntry = false;
          metadataAware = true;
       } else if (ice instanceof MetadataTransientCacheEntry) {
-         mortalEntry = false;
-         transientEntry = true;
          metadataAware = true;
       } else if (ice instanceof MetadataTransientMortalCacheEntry) {
-         mortalEntry = true;
-         transientEntry = true;
          metadataAware = true;
       } else {
-         mortalEntry = false;
-         transientEntry = false;
          metadataAware = false;
       }
+      Metadata metadata;
       if (metadataAware) {
-         // Assume it has a pointer for the metadata
-         iceSize += POINTER_SIZE;
-         // The metadata has itself and the class reference
-         metadataSize += OBJECT_SIZE + POINTER_SIZE;
-         Metadata metadata = ice.getMetadata();
-         if (metadata instanceof EmbeddedMetadata) {
-            // The embedded metadata has a reference and NumericVersion instance
+         metadata = ice.getMetadata();
+         // We don't support other metadata types currently
+         if (!(metadata instanceof EmbeddedMetadata)) {
+            metadata = null;
+         }
+      } else {
+         metadata = null;
+      }
+      long keyValueMetadataSize = calculateSize(key, ice.getValue(), metadata);
+      return keyValueMetadataSize + noMetadataSize;
+   }
+
+   @Override
+   public long calculateSize(K key, V value, Metadata metadata) {
+      long objSize = calculator.calculateSize(key, value);
```

Comment:
I was just wondering if we could hire a tool (or just copy some bits from it) that will calculate the size for us? Maybe this: http://openjdk.java.net/projects/code-tools/jol/?

Reply:
This could definitely be a different implementation of a SizeCalculator. At the time I used reflection and simple addition as it was substantially faster. Unfortunately this approach only supports primitives, primitive Wrappers, Strings and arrays.

One thing about jol is that it doesn't count nested objects. So this means we would have to invoke it at least 3 times (possibly 6) to estimate a CacheEntry (CacheEntry itself, key, value, Metadata, Version, Custom Metadata). This doesn't even take into account if the key or value have nested objects, which jol won't add in either.  Needless to say to support jol to be a valid estimation we would have to expose the Calculator to the user to implement and then tell us how to estimate use jol for their key and values as well.

Reply:
Actually I might be wrong about the nested part, I will have to look into this at a later point.